migrator/migrator.go

```diff
@@ -404,21 +403,25 @@ func (m Migrator) RenameColumn(value interface{}, oldName, newName string) error
 // MigrateColumn migrate column
 func (m Migrator) MigrateColumn(value interface{}, field *schema.Field, columnType gorm.ColumnType) error {
 	// found, smart migrate
-	fullDataType := strings.ToLower(m.DB.Migrator().FullDataTypeOf(field).SQL)
+	fullDataType := strings.TrimSpace(strings.ToLower(m.DB.Migrator().FullDataTypeOf(field).SQL))
 	realDataType := strings.ToLower(columnType.DatabaseTypeName())
 
 	alterColumn := false
 
+	// check type
+	if !field.PrimaryKey && !strings.HasPrefix(fullDataType, realDataType) {
+		alterColumn = true
+	}
+
 	// check size
 	if length, ok := columnType.Length(); length != int64(field.Size) {
 		if length > 0 && field.Size > 0 {
 			alterColumn = true
 		} else {
 			// has size in data type and not equal
 			// Since the following code is frequently called in the for loop, reg optimization is needed here
-			matches := regRealDataType.FindAllStringSubmatch(realDataType, -1)
```

Comment:
Remove `regRealDataType` because it is not match anything after https://github.com/go-gorm/gorm/pull/5088